File: test/language/array_types/cpp/PackedArraysMappingTest.cpp
#include "gtest/gtest.h"

#include "array_types/packed_arrays_mapping/PackedArraysMapping.h"

#include "zserio/RebindAlloc.h"

namespace array_types
{
namespace packed_arrays_mapping
{

using allocator_type = PackedArraysMapping::allocator_type;
using string_type = zserio::string<zserio::RebindAlloc<allocator_type, char>>;
template <typename T>
using vector_type = std::vector<T, zserio::RebindAlloc<allocator_type, T>>;

using BitBuffer = zserio::BasicBitBuffer<zserio::RebindAlloc<allocator_type, uint8_t>>;

class PackedArraysMappingTest : public ::testing::Test
{
protected:
    static const size_t fixedArrayLength = 5;
};

TEST_F(PackedArraysMappingTest, unsignedIntegerArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setUint8Array(vector_type<uint8_t>(fixedArrayLength));
    packedArraysMapping.setUint16Array(vector_type<uint16_t>(fixedArrayLength));
    packedArraysMapping.setUint32Array(vector_type<uint32_t>(fixedArrayLength));
    packedArraysMapping.setUint64Array(vector_type<uint64_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, signedIntegerArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setInt8Array(vector_type<int8_t>(fixedArrayLength));
    packedArraysMapping.setInt16Array(vector_type<int16_t>(fixedArrayLength));
    packedArraysMapping.setInt32Array(vector_type<int32_t>(fixedArrayLength));
    packedArraysMapping.setInt64Array(vector_type<int64_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, unsignedBitfieldArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setBitfield8Array(vector_type<uint8_t>(fixedArrayLength));
    packedArraysMapping.setBitfield16Array(vector_type<uint16_t>(fixedArrayLength));
    packedArraysMapping.setBitfield32Array(vector_type<uint32_t>(fixedArrayLength));
    packedArraysMapping.setBitfield63Array(vector_type<uint64_t>(fixedArrayLength));
    packedArraysMapping.setUint8Value(8);
    packedArraysMapping.setVariableBitfieldLongArray(vector_type<uint64_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, signedBitfieldArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setIntfield8Array(vector_type<int8_t>(fixedArrayLength));
    packedArraysMapping.setIntfield16Array(vector_type<int16_t>(fixedArrayLength));
    packedArraysMapping.setIntfield32Array(vector_type<int32_t>(fixedArrayLength));
    packedArraysMapping.setIntfield64Array(vector_type<int64_t>(fixedArrayLength));
    packedArraysMapping.setUint8Value(8);
    packedArraysMapping.setVariableIntfieldLongArray(vector_type<int64_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, variableUnsignedIntegerArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setVaruint16Array(vector_type<uint16_t>(fixedArrayLength));
    packedArraysMapping.setVaruint32Array(vector_type<uint32_t>(fixedArrayLength));
    packedArraysMapping.setVaruint64Array(vector_type<uint64_t>(fixedArrayLength));
    packedArraysMapping.setVaruintArray(vector_type<uint64_t>(fixedArrayLength));
    packedArraysMapping.setVarsizeArray(vector_type<uint32_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, variableSignedIntegerArrays)
{
    PackedArraysMapping packedArraysMapping;

    packedArraysMapping.setVarint16Array(vector_type<int16_t>(fixedArrayLength));
    packedArraysMapping.setVarint32Array(vector_type<int32_t>(fixedArrayLength));
    packedArraysMapping.setVarint64Array(vector_type<int64_t>(fixedArrayLength));
    packedArraysMapping.setVarintArray(vector_type<int64_t>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, compoundArray)
{
    PackedArraysMapping packedArraysMapping;
    packedArraysMapping.setCompoundArray(vector_type<TestStructure>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, enumArray)
{
    PackedArraysMapping packedArraysMapping;
    packedArraysMapping.setEnumArray(vector_type<TestEnum>(fixedArrayLength));
}

TEST_F(PackedArraysMappingTest, bitmaskArray)
{
    PackedArraysMapping packedArraysMapping;
    packedArraysMapping.setBitmaskArray(vector_type<TestBitmask>(fixedArrayLength));
}

} // namespace packed_arrays_mapping
} // namespace array_types

File: test/language/array_types/java/array_types/PackedArraysMappingTest.java
package array_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.math.BigInteger;
import java.util.Arrays;

import array_types.packed_arrays_mapping.PackedArraysMapping;
import array_types.packed_arrays_mapping.TestStructure;
import array_types.packed_arrays_mapping.TestEnum;
import array_types.packed_arrays_mapping.TestBitmask;

public class PackedArraysMappingTest
{
    @Test
    public void unsignedIntegerArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setUint8Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setUint16Array(new int[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setUint32Array(new long[FIXED_ARRAY_LENGTH]);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final BigInteger[] bigIntegerArray = new BigInteger[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setUint64Array(bigIntegerArray);
        assertTrue(Arrays.equals(bigIntegerArray, packedArraysMapping.getUint64Array()));
    }

    @Test
    public void signedIntegerArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setInt8Array(new byte[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setInt16Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setInt32Array(new int[FIXED_ARRAY_LENGTH]);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final long[] longArray = new long[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setInt64Array(longArray);
        assertTrue(Arrays.equals(longArray, packedArraysMapping.getInt64Array()));
    }

    @Test
    public void unsignedBitfieldArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setBitfield8Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setBitfield16Array(new int[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setBitfield32Array(new long[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setBitfield63Array(new long[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setUint8Value((short)8);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final BigInteger[] unsignedLongArray = new BigInteger[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setVariableBitfieldLongArray(unsignedLongArray);
        assertTrue(Arrays.equals(unsignedLongArray, packedArraysMapping.getVariableBitfieldLongArray()));
    }

    @Test
    public void signedBitfieldArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setIntfield8Array(new byte[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setIntfield16Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setIntfield32Array(new int[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setIntfield64Array(new long[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setUint8Value((short)8);
        packedArraysMapping.setVariableIntfieldLongArray(new long[FIXED_ARRAY_LENGTH]);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final long[] longArray = new long[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setVariableIntfieldLongArray(longArray);
        assertTrue(Arrays.equals(longArray, packedArraysMapping.getVariableIntfieldLongArray()));
    }

    @Test
    public void variableUnsignedIntegerArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setVaruint16Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setVaruint32Array(new int[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setVaruint64Array(new long[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setVaruintArray(new BigInteger[FIXED_ARRAY_LENGTH]);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final int[] varSizeArray = new int[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setVarsizeArray(varSizeArray);
        assertTrue(Arrays.equals(varSizeArray, packedArraysMapping.getVarsizeArray()));
    }

    @Test
    public void variableSignedIntegerArrays()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        packedArraysMapping.setVarint16Array(new short[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setVarint32Array(new int[FIXED_ARRAY_LENGTH]);
        packedArraysMapping.setVarint64Array(new long[FIXED_ARRAY_LENGTH]);

        // just do something with packedArraysMapping not to have SpotBugs warning
        final long[] varIntArray = new long[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setVarintArray(varIntArray);
        assertTrue(Arrays.equals(varIntArray, packedArraysMapping.getVarintArray()));
    }

    @Test
    public void compoundArray()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        // just do something with packedArraysMapping not to have SpotBugs warning
        final TestStructure[] objectArray = new TestStructure[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setCompoundArray(objectArray);
        assertTrue(Arrays.equals(objectArray, packedArraysMapping.getCompoundArray()));
    }

    @Test
    public void enumArray()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        // just do something with packedArraysMapping not to have SpotBugs warning
        final TestEnum[] enumArray = new TestEnum[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setEnumArray(enumArray);
        assertTrue(Arrays.equals(enumArray, packedArraysMapping.getEnumArray()));
    }

    @Test
    public void bitmaskArray()
    {
        PackedArraysMapping packedArraysMapping = new PackedArraysMapping();

        // just do something with packedArraysMapping not to have FindBugs warning
        final TestBitmask[] bitmaskArray = new TestBitmask[FIXED_ARRAY_LENGTH];
        packedArraysMapping.setBitmaskArray(bitmaskArray);
        assertTrue(Arrays.equals(bitmaskArray, packedArraysMapping.getBitmaskArray()));
    }

    private static final int FIXED_ARRAY_LENGTH = 5;
}
